### PR TITLE
Notifications initial implementation

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

WIP, waiting for comments. Initial try with notification system

Screenshot 2023-12-15 at 16 13 34

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by drodil** [comment]
Relates to #20312 #10652 and #639

Waiting for #21728 to get merged to get this work properly

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
Would also like to get @Rugvip to check this out, when you have time!

**Comment by drodil** [comment]
Re-open please

File: plugins/signals-react/src/api/SignalApi.ts
Comment (drodil) [inline plugins/signals-react/src/api/SignalApi.ts:25]: Moving these changes to own PR as these are signals related

Comment (drodil) [reply plugins/signals-react/src/api/SignalApi.ts:25]: Here #22529

File: beps/0001-notifications-system/README.md
Comment (Rugvip) [inline beps/0001-notifications-system/README.md:1]: Please open a separate PR for these changes

Comment (drodil) [reply beps/0001-notifications-system/README.md:1]: Here you go https://github.com/backstage/backstage/pull/22641

File: plugins/notifications-backend/src/service/router.ts
Comment (Rugvip) [inline plugins/notifications-backend/src/service/router.ts:357]: General note: never use `.send` unless that is exactly what you need and be careful when using it. It easily introduces XSS vulnerabilities. In this case `.json` instead.

File: plugins/notifications-node/src/lib.ts
Comment (Rugvip) [inline plugins/notifications-node/src/lib.ts:45]: Unused right?

File: plugins/notifications-node/src/service/DefaultNotificationService.ts
Comment (Rugvip) [inline plugins/notifications-node/src/service/DefaultNotificationService.ts:83]: Noting that we'll want this to be based on the service auth instead. Could you add a `// TODO` for that for now?

File: plugins/notifications-node/src/service/DefaultNotificationService.ts
Comment (Rugvip) [inline plugins/notifications-node/src/service/DefaultNotificationService.ts:91]: Could we throw for now until we settle exactly what we want failures to look like? It's the less surprising option. I do think we'll want some form of feature discovery in place so that we can try to post notifications conditionally btw. Some work going on towards that in #22671

File: plugins/notifications-node/src/service/NotificationService.ts
Comment (Rugvip) [inline plugins/notifications-node/src/service/NotificationService.ts:21]: It should already be instantiated for each plugin directly, best skip this step, regardless of whether it's for the old or new backend system.

File: plugins/notifications/config.d.ts
Comment (Rugvip) [inline plugins/notifications/config.d.ts:29]: With the new frontend system in mind we'll most likely move away from this kind of global config. It'd be possible to just leave this as props for the sidebar item for now right?

File: plugins/notifications/src/api/NotificationsApi.ts
Comment (Rugvip) [inline plugins/notifications/src/api/NotificationsApi.ts:50]: Thinking it's best to keep this one as batching of individual updates, rather that the same update applied to a batch. i.e. `updateNotifications(updates: NotificationUpdate[])` potentially wrapping it up more than that though.

File: plugins/notifications/src/hooks/useTitleCounter.ts
Comment (Rugvip) [inline plugins/notifications/src/hooks/useTitleCounter.ts:29]: Just a note that we clearly need something more specific in place to handle titles from a framework perspective 😅

File: plugins/signals-react/src/api/SignalApi.ts
Comment (Rugvip) [inline plugins/signals-react/src/api/SignalApi.ts:29]: This is a merge conflict thing right?

**Review by Rugvip — CHANGES_REQUESTED** [review]
Sweet, thank you! 🎉 

Things I think would best to fix before merge:
- Remove support for the old backend system
- Remove config from frontend plugin
- Remove the signal plugin rollbacks

File: plugins/notifications-backend/migrations/20231215_init.js
Comment (drodil) [reply plugins/notifications-backend/migrations/20231215_init.js:22]: Except for the search in notifications table as presented in the other implementation? Do we want it?

File: plugins/notifications-backend/migrations/20231215_init.js
Comment (drodil) [reply plugins/notifications-backend/migrations/20231215_init.js:18]: Yeah, maybe makes sense. Need some refactoring though

Comment (drodil) [reply plugins/notifications-backend/migrations/20231215_init.js:18]: After thinking about it, I bet having a separate table for user interactions adds unnecessary complexity. Usually, there's only one column you update at a time (read, done, saved) and the rest of the notification stays as is. The only case when there is a need to update other metadata of the notification is when using the `scope` to update an existing notification. And in this case, it also just updates one row. 

I would say we merge like this and talk more if we need to make this more complex later, what do you think?

File: plugins/notifications/src/api/NotificationsApi.ts
Comment (drodil) [reply plugins/notifications/src/api/NotificationsApi.ts:50]: Thinking about the use case of updating multiple notifications, I think it makes more sense to keep it as is. The notifications are shown by their status and are updated either in batch or as a single notification always and with the same parameters (set as done, set as undone, save etc.). We can modify this later if use case for changing notifications somehow differently rises?

Comment (Rugvip) [reply plugins/notifications/src/api/NotificationsApi.ts:50]: Alright let's leave as is

File: plugins/notifications-backend/migrations/20231215_init.js
Comment (Rugvip) [reply plugins/notifications-backend/migrations/20231215_init.js]: Yeah we can certainly iterate in this bit

File: plugins/notifications-backend/migrations/20231215_init.js
Comment (Rugvip) [reply plugins/notifications-backend/migrations/20231215_init.js:22]: I'd imagine it'd be best to have an extracted search string in that case, but we can tweak this later too. Could be that this is fine and we do a separate metadata blob instead

File: plugins/notifications/package.json
Comment (Rugvip) [inline plugins/notifications/package.json:51]: The versions of the testing library deps and react are too old, could you align them with the rest of the repo?

File: packages/backend/package.json
Comment (Rugvip) [inline packages/backend/package.json:59]: Can skip these

File: plugins/notifications-backend/api-report.md
Comment (Rugvip) [inline plugins/notifications-backend/api-report.md:18]: Let's try to skip exporting it

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts
Comment (Rugvip) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:33]: Lot of things marked public here that aren't. Use `@internal` instead

**Review by Rugvip — CHANGES_REQUESTED** [review]
Yep just about ready to go, couple more nits I found on this pass

**Comment by Rugvip** [comment]
For running the new backend system there are a few `:next` scripts in the root package. For running it in a local plugin setup you use `dev/index.ts` as the entry point (outside src) and just set up a normal backend with the desired plugins.

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts
Comment (drodil) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:33]: Yeah, I change these. Probably the `NotificationsStore` interface should be public still?

Comment (Rugvip) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:33]: Yep we'd add an extension point for that in the future if it's something we feel makes sense. We've generally backtracked on the idea of public storage interfaces though and usually avoid it

**Review by Rugvip — APPROVED** [review]
Awesome, let's :shipit:! 😁

**Pull request merged**
